Add deallocator references

agibsonccc opened the pull request:
## What changes were proposed in this pull request?
ReferenceMetadata is added to deallocators .
This is to prevent deallocation of constant buffers.
This is also expandable to other metadata later
as needed to control allocation behavior.
Minor build updates (release version windows fixes)
Fixes windows  build issue
Rewrite shape descriptor allocation to use pointers.
Add kotlin-reflect to model import to ensure the new op annotation scanning works out of the box.

(Please fill in changes proposed in this fix)

## How was this patch tested?

Ran most recent tests to determine further source of crashes

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.konduit.ai/multi-project/how-to-guides/contribute/eclipse-contributors) page for details
- [X ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ X] Created tests for any significant new code additions.
- [ X] Relevant tests for your changes are passing.

agibsonccc committed:
Add new  reference metadata to cover when deallocators are dealing with constant buffers
Shorten PV concurrency test for automated tests

agibsonccc committed:
Add new  reference metadata to cover when deallocators are dealing with constant buffers
Shorten PV concurrency test for automated tests
Remove dll export from hidden function

agibsonccc committed:
Fix cbow inference segfaults
Fix scalar definition
Misc spacing fixes
Remove commented code
Remove extra print

agibsonccc committed:
Add new  reference metadata to cover when deallocators are dealing with constant buffers
Shorten PV concurrency test for automated tests
Remove dll export from hidden function
Remove tmate

agibsonccc committed:
Add new  reference metadata to cover when deallocators are dealing with constant buffers
Shorten PV concurrency test for automated tests
Remove dll export from hidden function
Remove tmate
Change default version release version to M3

treo commented on nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/memory/deallocation/DeallocatorService.java:
```diff
@@ -48,7 +48,7 @@ public class DeallocatorService {
         // we need to have at least 2 threads, but for CUDA we'd need at least numDevices threads, due to thread->device affinity
         int numDevices = Nd4j.getAffinityManager().getNumberOfDevices();
-        int numThreads = Math.max(2, numDevices * 2);
+        int numThreads = 1;
```
Comment and value don't fit anymore.

agibsonccc committed:
Add new  reference metadata to cover when deallocators are dealing with constant buffers
Shorten PV concurrency test for automated tests
Remove dll export from hidden function
Remove tmate
Change default version release version to M3

treo commented on libnd4j/include/array/NDArray.hXX:
```diff
@@ -1578,6 +1578,7 @@ void NDArray::reduceNumber(sd::reduce::LongOps op, NDArray &target, void *extraP
 NDArray NDArray::indexReduceNumber(sd::indexreduce::Ops op, ExtraArguments *extraParams) {
   if (isS()) throw std::runtime_error("NDArray::indexReduceNumber: you can't use this method on String array!");
 
+  sd_printf("indexReduceNumber before exec\n",0);
```
printf outside of error condition.

treo commented on libnd4j/include/array/NDArray.hXX:
```diff
@@ -1586,6 +1587,8 @@ NDArray NDArray::indexReduceNumber(sd::indexreduce::Ops op, ExtraArguments *extr
       res.specialBuffer(), res.specialShapeInfo());
   registerUse({&res}, {this});
+  sd_printf("After before exec\n",0);
+
```
printf outside of error condition.

treo commented on libnd4j/include/array/NDArray.hXX:
```diff
@@ -5199,8 +5202,16 @@ void NDArray::setShapeInfo(const sd::LongType *shapeInfo, const sd::DataType dty
+
 
+  sd_printf("Descriptor about to set: rank %d data type %d is empty %d\n" ,descriptor.rank(),descriptor.dataType(),descriptor.isEmpty());
+  sd_printf("Before obtaining constant shape info for descriptor\n",0);
```
printf outside of error condition.

treo commented on libnd4j/include/array/NDArray.hXX:
```diff
@@ -5199,8 +5202,16 @@ void NDArray::setShapeInfo(const sd::LongType *shapeInfo, const sd::DataType dty
+  sd_printf("Before obtaining constant shape info for descriptor\n",0);
+
+  auto shapeBuffer = ConstantShapeHelper::getInstance().bufferForShapeInfo(const_cast<ShapeDescriptor &>(descriptor));
+  sd_printf("After obtaining constant shape info for descriptor\n",0);
```
printf outside of error condition.

treo commented on libnd4j/include/array/impl/ShapeDescriptor.cpp:
```diff
@@ -153,11 +153,24 @@ ShapeDescriptor::ShapeDescriptor(const DataType type, const sd::LongType length)
   _extraProperties = ArrayOptions::propertyWithoutDataType(shapeInfo);
+  sd_printf("ShapeDescriptor constructor: Determined extra properties\n",0);
   if (inheritDtype) _dataType = ArrayOptions::dataType(shapeInfo);
+  sd_printf("ShapeDescriptor constructor: Determined dtype\n",0);
```
even more printf outside of error condition.

treo reviewed (changes requested):
I guess this is WIP again? A lot of printfs now. 

I've stopped commenting on them when I saw just how many you've added.

agibsonccc committed:
WIP with printing

agibsonccc committed:
WIP with printing

agibsonccc committed:
Add deletes for new descriptor creation

agibsonccc committed:
Rewrite shape desriptor allocation. Migrate to heap based pointers rather than stack based references to control
deallocation of references better.
Update external allocation counts to be correct

agibsonccc committed:
Rewrite shape desriptor allocation. Migrate to heap based pointers rather than stack based references to control
deallocation of references better.
Update external allocation counts to be correct
Fix kotlin-reflect usage by default for custom op annotation scanning when using model import

agibsonccc committed:
Fix up factory method deallocations

treo commented on libnd4j/include/array/impl/NDArrayFactory.cpp:
```diff
@@ -506,11 +512,13 @@ NDArray NDArrayFactory::create(const char order, const std::vector<sd::LongType>
+            "                               sd::LaunchContext* context)\n",0);
 
-  ShapeDescriptor descriptor(dtype, order, shape);
+  ShapeDescriptor *descriptor = new ShapeDescriptor(dtype, order, shape);
```
Missing a delete for this one?

treo commented on libnd4j/include/array/impl/NDArrayFactory.cpp:
```diff
@@ -635,17 +644,19 @@ NDArray* NDArrayFactory::create_(const char order, const std::vector<sd::LongTyp
 template <typename T>
 NDArray NDArrayFactory::create(T* buffer, const char order, const std::initializer_list<sd::LongType>& shape,
                                sd::LaunchContext* context) {
+  sd_printf("In arrLength usage at  NDArrayFactory::create(T* buffer, const char order, const std::initializer_list<sd::LongType>& shape,\n"
```
printf outside of error condition.

treo commented on libnd4j/include/ops/declarable/generic/nlp/cbow.cpp:
```diff
@@ -92,22 +93,26 @@ CONFIGURABLE_OP_IMPL(cbow_inference, 6, 6, true, -2, -2) {
   auto lockedWordsArr = new NDArray(lockedWordsOne);
+  sd_printf("After words array creation\n",0);
 
+  sd_printf("After array creation\n",0);
```
printf's outside of error conditions.

agibsonccc committed:
Fix up factory method deallocations

agibsonccc committed:
Fix up factory method deallocations
Remove extra printfs
Add 1 more deallocation

agibsonccc committed:
Fix up factory method deallocations
Remove extra printfs
Add 1 more deallocation
Revert deallocator service threads change
